drivers/can/can_mcp25xxfd.c

```diff
@@ -0,0 +1,1084 @@
+	cmd_buf[1] = address & 0xFF;
+	const struct spi_buf tx_buf[] = {
+		{ .buf = cmd_buf, .len = sizeof(cmd_buf) },
+	};
+	const struct spi_buf rx_buf[] = {
+		{ .buf = cmd_buf, .len = sizeof(cmd_buf) },
+	};
+	const struct spi_buf_set tx = { .buffers = tx_buf,
+					.count = ARRAY_SIZE(tx_buf) };
+	const struct spi_buf_set rx = { .buffers = rx_buf,
+					.count = ARRAY_SIZE(rx_buf) };
+	int ret = spi_transceive(DEV_DATA(dev)->spi, &DEV_DATA(dev)->spi_cfg,
+				 &tx, &rx);
+	memcpy(rxd, &cmd_buf[2], rx_len);
+	return ret;
+}
+
+static int mcp25xxfd_write(const struct device *dev, uint16_t address,
+			   void *txd, uint8_t tx_len)
+{
+	uint8_t cmd_buf[] = { (MCP25XXFD_OPCODE_WRITE << 4) +
+			      ((address >> 8) & 0xF),
+			      address & 0xFF };
+	const struct spi_buf tx_buf[] = {
+		{ .buf = cmd_buf, .len = sizeof(cmd_buf) },
+		{ .buf = txd, .len = tx_len },
+	};
+	const struct spi_buf_set tx = { .buffers = tx_buf,
+					.count = ARRAY_SIZE(tx_buf) };
+
+	return spi_write(DEV_DATA(dev)->spi, &DEV_DATA(dev)->spi_cfg, &tx);
+}
+
+static int mcp25xxfd_fifo_read(const struct device *dev, uint16_t fifo_address,
+			       void *rxd, uint8_t rx_len)
+{
+	struct mcp25xxfd_data *dev_data = DEV_DATA(dev);
+	union mcp25xxfd_fifo fiforegs;
+	int ret;
+
+	k_mutex_lock(&dev_data->mutex, K_FOREVER);
+
+	ret = mcp25xxfd_read(dev, fifo_address, &fiforegs, sizeof(fiforegs));
+	if (ret < 0) {
+		goto done;
+	}
+
+	if (!fiforegs.sta.FNEIF) {
+		ret = -ENODATA;
+		goto done;
+	}
+
+	ret = mcp25xxfd_read(dev, 0x400 + fiforegs.ua, rxd, rx_len);
+	if (ret < 0) {
+		goto done;
+	}
+
+	fiforegs.con.UINC = 1;
+	ret = mcp25xxfd_writeb(dev, fifo_address + 1, &fiforegs.con.bytes[1]);
+
+done:
+	k_mutex_unlock(&dev_data->mutex);
+	return ret;
+}
+
+static int mcp25xxfd_fifo_write(const struct device *dev, uint16_t fifo_address,
+				void *txd, uint8_t tx_len)
+{
+	struct mcp25xxfd_data *dev_data = DEV_DATA(dev);
+	union mcp25xxfd_fifo fiforegs;
+	int ret;
+
+	k_mutex_lock(&dev_data->mutex, K_FOREVER);
+
+	ret = mcp25xxfd_read(dev, fifo_address, &fiforegs, sizeof(fiforegs));
+	if (ret < 0) {
+		goto done;
+	}
+
+	if (!fiforegs.sta.FNEIF) {
+		ret = -ENOMEM;
+		goto done;
+	}
+
+	ret = mcp25xxfd_write(dev, 0x400 + fiforegs.ua, txd, tx_len);
+	if (ret < 0) {
+		goto done;
+	}
+
+	fiforegs.con.UINC = 1;
+	fiforegs.con.TXREQ = 1;
+	ret = mcp25xxfd_writeb(dev, fifo_address + 1, &fiforegs.con.bytes[1]);
+
+done:
+	k_mutex_unlock(&dev_data->mutex);
+	return ret;
+}
+
+static void mcp25xxfd_zcanframe_to_txobj(const struct zcan_frame *src,
+					 struct mcp25xxfd_txobj *dst)
+{
+	memset(dst, 0, sizeof(struct mcp25xxfd_txobj));
+	if (src->id_type == CAN_STANDARD_IDENTIFIER) {
+		dst->SID = src->id;
+	} else {
+		dst->SID = src->id >> 18;
+		dst->EID = src->id;
+		dst->IDE = 1;
+	}
+	dst->BRS = src->brs;
+	dst->RTR = src->rtr == CAN_REMOTEREQUEST;
+	dst->DLC = src->dlc;
+#if defined(CONFIG_CAN_FD_MODE)
+	dst->FDF = src->fd;
+#endif
+	memcpy(dst->DATA, src->data, CAN_MAX_DLEN);
```

Comment:
memcpy(dst->data, src->DATA, src->dlc); ?

Because dst is 
uint8_t objbuf[sizeof(struct mcp25xxfd_txobj) + msg->dlc];